libraries/Microsoft.Bot.Builder/TokenResolver.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder
+{
+    internal class TokenResolver
+    {
+        private const int PollingInterval = 1000;   // Poll for token every 1 second.
+
+        public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null)
+            {
+                foreach (var attachment in activity?.Attachments?.Where(a => a.ContentType == OAuthCard.ContentType))
+                {
+                    OAuthCard oauthCard = attachment.Content as OAuthCard;
+                    if (oauthCard != null && !string.IsNullOrEmpty(oauthCard.ConnectionName))
+                    {
+                        // Poll as a background task
+                        Task.Run(() => PollForTokenAsync(adapter, turnContext, activity, oauthCard.ConnectionName, cancellationToken));
+                    }
+                }
+            }
+        }
+
+        private static async Task PollForTokenAsync(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, string connectionName, CancellationToken cancellationToken)
+        {
+            TokenResponse tokenResponse = null;
```

Comment:
Exception leakage here, based on the Task.run, doesn't seem correct.

Reply:
I added a continuation step.